site/src/job_queue.rs

```diff
@@ -36,6 +41,93 @@ async fn create_benchmark_request_master_commits(
     Ok(())
 }
 
+/// Parses strings in the following formats extracting the Date & release tag
+/// `static.rust-lang.org/dist/2016-05-24/channel-rust-1.9.0.toml`
+/// `static.rust-lang.org/dist/2016-05-31/channel-rust-nightly.toml`
+/// `static.rust-lang.org/dist/2016-06-01/channel-rust-beta.toml`
+/// `static.rust-lang.org/dist/2025-06-26/channel-rust-1.89-beta.toml`
+/// `static.rust-lang.org/dist/2025-06-26/channel-rust-1.89.0-beta.toml`
+/// `static.rust-lang.org/dist/2025-06-26/channel-rust-1.89.0-beta.2.toml`
+fn parse_release_string(url: &str) -> anyhow::Result<Option<(String, DateTime<Utc>)>> {
+    static VERSION_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"(\d+\.\d+\.\d+)").unwrap());
+
+    // Grab ".../YYYY-MM-DD/FILE.toml" components with Path helpers.
+    let file = Path::new(url)
+        .file_name()
+        .and_then(|n| n.to_str())
+        .ok_or_else(|| anyhow::anyhow!("URL lacks a file name"))?;
+
+    let date_str = Path::new(url)
+        .parent()
+        .and_then(Path::file_name)
+        .and_then(|n| n.to_str())
+        .ok_or_else(|| anyhow::anyhow!("URL lacks a date segment"))?;
+
+    // No other beta releases are recognized as toolchains.
+    //
+    // We also have names like this:
+    //
+    // * channel-rust-1.75-beta.toml
+    // * channel-rust-1.75.0-beta.toml
+    // * channel-rust-1.75.0-beta.1.toml
+    //
+    // Which should get ignored for now, they're not consumable via rustup yet.
+    if file.contains("beta") && file != "channel-rust-beta.toml" {
+        return Ok(None);
+    }
+
+    // Parse the YYYY-MM-DD segment and stamp it with *current* UTC time.
+    let naive = NaiveDate::parse_from_str(date_str, "%Y-%m-%d")?;
+    let now = Utc::now();
+    let published = naive
+        .and_hms_nano_opt(now.hour(), now.minute(), now.second(), now.nanosecond())
+        .expect("valid HMS")
+        .and_local_timezone(Utc)
+        .single()
+        .unwrap();
+
+    // Special-case the rolling beta channel.
+    if file == "channel-rust-beta.toml" {
+        return Ok(Some((format!("beta-{date_str}"), published)));
+    }
+
+    // Otherwise pull out a semver like "1.70.0" and return it.
+    if let Some(cap) = VERSION_RE.captures(file).and_then(|m| m.get(1)) {
+        return Ok(Some((cap.as_str().to_owned(), published)));
+    }
+
+    Ok(None)
+}
+
+/// Store the latest release commits or do nothing if all of them are
+/// already in the database
+async fn create_benchmark_request_releases(
+    conn: &dyn database::pool::Connection,
+) -> anyhow::Result<()> {
+    let releases: String = reqwest::get("https://static.rust-lang.org/manifests.txt")
+        .await?
+        .text()
+        .await?;
+    // TODO; delete at some point in the future
+    let cutoff: chrono::DateTime<Utc> = chrono::DateTime::from_str("2025-06-01T00:00:00.000Z")?;
```

Comment:
The original code loaded 20 most recent artifacts, I think that we can do the same here. You can still keep the cut-off, but we should also have some long-term cap on the number of release artifacts parsed. You can replicate the original code, so iterate through lines in reverse order, try to parse release artifacts, and take the first 20 of them.